File: src/main/js/formulas/math/LOG10.ts
import { NumException } from "../../errors/NumException";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class LOG10 extends AbstractFormula {
  static SELF: LOG10 = new LOG10();
  NAME = FormulaName.LOG10;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    const value = Converters.toNumber(Converters.first(this.collateralLookup(origin, values[0])));
    if (value <= 0) {
      return new NumException(`LOG10 parameter 1 value is ${value}. It should be greater than 0.`);
    }
    return Math["log10"](value);
  }
}

File: src/main/java/io/protobase/f7/formulas/math/LOG10.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NumException;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class LOG10 extends AbstractFormula {
  public static FormulaName NAME = FormulaName.LOG10;
  public static LOG10 SELF = new LOG10();

  public LOG10() {
    super();
  }

  public LOG10(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    Double value = Converters.toDouble(Converters.first(collateralLookup.apply(origin, values[0])));
    if (value <= 0) {
      return new NumException(String.format("LOG10 parameter 1 value is %f. It should be greater than 0.", value));
    }
    return Math.log10(value);
  }
}
